src/audio_input/audio.c

```diff
@@ -0,0 +1,186 @@
+   SDL_EventState(SDL_AUDIODEVICEREMOVED, SDL_ENABLE);
+   if (pthread_spin_init(&audio->sampleGetLock, PTHREAD_PROCESS_PRIVATE)) {
+      rsError("pthread_spin_init failed");
+   }
+
+   SDL_zero(audio->ispec);
+   audio->ispec.freq = (int)config->audioSamplerate;
+   audio->ispec.format = AUDIO_S16LSB;
+   audio->ispec.channels = (Uint8)config->audioChannels;
+   audio->ispec.samples = SAMPLES_PER_CALLBACK; // must be power of 2
+   audio->ispec.callback = audioCallback;
+   audio->ispec.userdata = audio;
+
+   probeDevices(audio);
+   audio->deviceName = getDeviceName(config->audioDeviceName);
+   if (!openDevice(audio, audio->deviceName)) {
+      rsError("Exiting");
+   }
+
+   int size1s = audio->ospec.channels * audio->ospec.freq * (int)sizeof(uint16_t);
+   int sizeTotal = size1s * config->duration;
+   int numOfCallbacks =
+       (sizeTotal + (SAMPLES_PER_CALLBACK - 1)) / SAMPLES_PER_CALLBACK; // round up div
+   sizeTotal = numOfCallbacks * SAMPLES_PER_CALLBACK;
+   audio->sizeBatch = size1s / config->framerate;
+   rsCircleStaticCreate(&audio->data, (size_t)sizeTotal);
+   audio->deviceAddTime = 0;
+   audio->deviceRemoveTime = 0;
+   SDL_PauseAudioDevice(audio->deviceId, 0);
+}
+
+static void rsAudioReconnect(RSAudio *audio, const char *devname) {
+   SDL_CloseAudioDevice(audio->deviceId);
+   /*
+    * set the audio buffor index to 0. Effectively erasing all
+    * the collected samples so far. This ensures that we'll be
+    * back in sync after the audio setup is complete.
+    * Once the frame pace and audio timestamps are implemented
+    * it should be calculated how long the audio setup took
+    * and move the index accordingly
+    */
+   audio->data.index = 0;
+   probeDevices(audio);
+   char *newname = getDeviceName(devname);
+   bool ok = openDevice(audio, newname);
+   rsMemoryDestroy(newname);
+   if (!ok) {
+      // just try opening the NULL device
+      if (!openDevice(audio, 0)) {
+         rsLog("Couldn't open any audio device. You will have no sound");
+         return;
+      }
+   }
+   SDL_PauseAudioDevice(audio->deviceId, 0);
+}
+
+void rsAudioGetSamples(RSAudio *audio, uint8_t *newbuff, int rewindFrames) {
+   size_t oldIndex = audio->data.index;
+   pthread_spin_lock(&audio->sampleGetLock);
+   rsCircleStaticMoveBackIndex(&audio->data, (size_t)(rewindFrames * audio->sizeBatch));
+   rsCircleStaticGet(&audio->data, newbuff, audio->data.size);
+   audio->data.index = oldIndex;
+   pthread_spin_unlock(&audio->sampleGetLock);
+}
+
+void rsAudioHandleEvents(RSAudio *audio) {
+   if (audio->deviceAddTime && (SDL_GetTicks() > audio->deviceAddTime)) {
+      rsLog("New audio devices detected. Attempting reconnect");
+      rsAudioReconnect(audio, audio->deviceName);
+      audio->deviceAddTime = 0;
+   }
+   if (audio->deviceRemoveTime && (SDL_GetTicks() > audio->deviceRemoveTime)) {
+      rsLog("Audio device disconnected. Trying to connect to another device");
+      rsAudioReconnect(audio, "default");
```

Comment:
The RemovedDevice event only comes when the currently device is disconnected. So if you only unplog your microphone or whatever else you are not using nothing is gonna happen.

Reply:
Still a good idea to go through the usual connect flow instead of resetting to `default`. I would personally want it to try and connect to one of my other output devices instead of my input devices.

Reply:
Wait I just remembered that `default` is output right? I still disagree with the naming but I'll make a comment about that in a sec.